frontend/src/lib/hostComm/HostCommunicationManager.tsx

```diff
@@ -0,0 +1,246 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ICustomThemeConfig, WidgetStates } from "src/lib/proto"
+
+import {
+  IAllowedMessageOriginsResponse,
+  IGuestToHostMessage,
+  IHostToGuestMessage,
+  VersionedMessage,
+  IMenuItem,
+  IToolbarItem,
+  DeployedAppMetadata,
+} from "./types"
+import { isValidOrigin } from "src/lib/util/UriUtil"
+// Uncomment this import to test host communication with
+// frontend/hostframe.html:
+import { IS_DEV_ENV } from "src/lib/baseconsts"
+
+import Resolver from "src/lib/util/Resolver"
+
+export const HOST_COMM_VERSION = 1
+
+export interface HostCommunicationProps {
+  readonly theme: {
+    setImportedTheme: (themeInfo: ICustomThemeConfig) => void
+  }
+  readonly sendRerunBackMsg: (
+    widgetStates?: WidgetStates,
+    pageScriptHash?: string
+  ) => void
+  readonly onPageChange: (pageScriptHash: string) => void
+  readonly closeModal: () => void
+  readonly stopScript: () => void
+  readonly rerunScript: () => void
+  readonly clearCache: () => void
+  readonly isOwnerChanged: (isOwner: boolean) => void
+  readonly hostMenuItemsChanged: (menuItems: IMenuItem[]) => void
+  readonly hostToolbarItemsChanged: (toolbarItems: IToolbarItem[]) => void
+  readonly hostHideSidebarNavChanged: (hideSidebarNav: boolean) => void
+  readonly sidebarChevronDownshiftChanged: (
+    sidebarChevronDownshift: number
+  ) => void
+  readonly pageLinkBaseUrlChanged: (pageLinkBaseUrl: string) => void
+  readonly queryParamsChanged: (queryParams: string) => void
+  readonly deployedAppMetadataChanged: (
+    deployedAppMetadata: DeployedAppMetadata
+  ) => void
+}
+
+/**
+ * Manages host communication & messaging
+ */
+export default class HostCommunicationManager {
+  public readonly props: HostCommunicationProps
+
+  private allowedOrigins: string[]
+
+  public deferredAuthToken: Resolver<string | undefined>
```

Comment:
Is this intended to be publicly writeable? If not, it should be `private`, and exposed in a public getter so that external code can't mutate it.